src/scene/glsl/graphics_mesh.frag

```diff
@@ -141,80 +142,45 @@ void main()
     // if (in_clip < -eps)
     //     discard;
 
-    vec3 normal, light_dir, light_color, ambient, diffuse, view_dir, reflect_dir, specular, color;
-    vec4 lpar;
-    vec3 lpos;
-    float diff, spec, view_facing;
-
     // Stroke parameters.
     float linewidth = params.linewidth;
     vec3 edgecolor = params.edgecolor.rgb;
     float edge_alpha = params.edgecolor.a;
     vec3 pos_tr;
 
-    normal = normalize(in_normal);
-    out_color = vec4(0, 0, 0, 1);
-    diffuse = vec3(0);
-    specular = vec3(0);
+    vec3 normal = normalize(in_normal);
 
     // Texture.
+    vec4 color;
     if (MESH_TEXTURED > 0)
     {
         // in this case, in_uvcolor.xy is uv coordinates
-        color = texture(tex, in_uvcolor.xy).xyz;
+        color = texture(tex, in_uvcolor.xy);
     }
     // Color.
     else
     {
-        color = in_uvcolor.xyz; // rgb
+        color = in_uvcolor; // rgba
     }
 
     // Lighting.
     if (MESH_LIGHTING > 0)
     {
-        pos_tr = ((mvp.model * vec4(in_pos, 1.0))).xyz;
-        view_dir = normalize(-mvp.view[3].xyz - pos_tr);
+        vec4 cam_pos = inverse(mvp.view) * vec4(0.0, 0.0, 0.0, 1.0);  // Better in vertex shader?
+
+        mat4 light_pos = params.light_dir;
+        mat4 light_color = params.light_color;
+        mat4 material = params.light_params;
 
-        for (int i = 0; i < 4; i++)
+        // TODO:  Update visual and panel setter functions.
+        // Current visual setters don't set w and a.
+        for (int i=0; i<4; i++)
```

Comment:
That will be fixed directly in the C functions that set the light and material parameters, they'll use the `dvz_params*()` functions with the appropriate values.